common/Dockerfile

```diff
@@ -0,0 +1,7 @@
+FROM busybox
+
+RUN wget -O /dumb-init https://github.com/Yelp/dumb-init/releases/download/v1.2.1/dumb-init_1.2.1_amd64
+RUN chmod +x /dumb-init
```

Comment:
Hva var greia med dumb-init igjen?

Reply:
https://github.com/Yelp/dumb-init

Fikser masse Docker-"bugs" (ikke bugs akkurat, men bivirkninger av at ting kjører i en container som PID 1)

Reply:
Hehe, ser jeg er for sein nå, men likevel: er baseimagene våre faktisk eksponert for disse problemene eller er det for å være på den trygge siden fordi vi ikke har helt kontroll at dette brukes?

Reply:
Hvis det hadde vært snakk om images som spawner massevis av prosesser som kan fullføre/dø individuelt gir det mening å bruke dette, men våre baseimages starter vel bare en prosess (jvm/node) og så lenge signaler forwardes til disse fra shellet som kjører som PID 1 er vel dette ikke et problem? eller?

Reply:
Du skulle kanskje ikke tro det... men Node.js er skikkelig dårlig til å kjøre som PID 1. For eksempel så håndterer den ikke signaler (SIGINT/SIGTERM/SIGKILL) med mindre du eksplisitt skriver dine egne signal handlers - noe de fleste ikke gjør.

Prøv å kjøre dette lokalt i terminalen på laptopen:

`node -e "require('http').createServer().listen(8080);"`

Den starter da en node-prosess. Hvis du trykker `Ctrl + C`, så sender du SIGINT til prosessen, som starter en gradeful shutdown.

I motsetning til hvis du prøver dette:

`docker run -it node node -e "require('http').createServer().listen(8080);"`

Den kjører akkurat det samme scriptet, men som PID 1 inne i en Docker-container. Hvis du nå prøver `Ctrl + C` for å stoppe containeren, så skjer det ingenting - siden den ikke håndterer signals. Du kan håndtere dem på egen hånd:

`docker run -it node node -e "require('http').createServer().listen(8080);process.on('SIGINT', () => process.exit(0))"`

Da funker `Ctrl + C`. I tillegg så må man håndtere alle de andre signalene.

[Node.js was not designed to run as PID 1 which leads to unexpected behaviour when running inside of Docker. For example, a Node.js process running as PID 1 will not respond to SIGTERM (CTRL-C) and similar signals.](https://github.com/nodejs/docker-node/blob/master/docs/BestPractices.md#handling-kernel-signals)

Reply:
Superenkel forskning viser at jvmen håndterer dette fint. Mulig at eldre jvmer ikke håndterer det like bra. https://github.com/jan-berge-ommedal/dumb-docker/tree/master/jvm-pid1

Men det er uansett en smal sak å kjøre node eller jvmen som noe annet enn pid 1 og likevel beholde signalhåndtering slik at man slipper å bruke binaries produsert av restaurantbransjen (som vel alltid er et skalkeskjul for annen luguber virksomhet :) ): https://github.com/jan-berge-ommedal/dumb-docker/tree/master/node-pid2

Reply:
JVMen hadde problemer med å lage heap dumps når den kjørte som PID 1. 
Se #2

Reply:
Alt er mulig, om du vil :P men det enkleste er å legge inn dumb-init i imaget, og så kan utviklerne leve lykkelig uvitende om at dette med PID 1 er et problem i det hele tatt!